src/main/java/io/supertokens/cronjobs/bulkimport/ProcessBulkImportUsers.java

```diff
@@ -0,0 +1,597 @@
+                appIdentifier.getAppId(), firstTenant);
+
+        if (lm.recipeId.equals("emailpassword")) {
+            processEmailPasswordLoginMethod(tenantIdentifier, storage, lm);
+        } else if (lm.recipeId.equals("thirdparty")) {
+            processThirdPartyLoginMethod(tenantIdentifier, storage, lm);
+        } else if (lm.recipeId.equals("passwordless")) {
+            processPasswordlessLoginMethod(tenantIdentifier, storage, lm);
+        } else {
+            throw new StorageTransactionLogicException(
+                    new IllegalArgumentException("Unknown recipeId " + lm.recipeId + " for loginMethod "));
+        }
+
+        associateUserToTenants(main, appIdentifier, storage, lm, firstTenant);
+    }
+
+    private void processEmailPasswordLoginMethod(TenantIdentifier tenantIdentifier, Storage storage,
+            LoginMethod lm) throws StorageTransactionLogicException {
+        try {
+            ImportUserResponse userInfo = EmailPassword.createUserWithPasswordHash(tenantIdentifier, storage, lm.email,
+                    lm.passwordHash, lm.timeJoinedInMSSinceEpoch);
+
+            lm.superTokensUserId = userInfo.user.getSupertokensUserId();
+        } catch (StorageQueryException | TenantOrAppNotFoundException e) {
+            throw new StorageTransactionLogicException(e);
+        } catch (DuplicateEmailException e) {
+            throw new StorageTransactionLogicException(
+                    new Exception("A user with email " + lm.email + " already exists"));
+        }
+    }
+
+    private void processThirdPartyLoginMethod(TenantIdentifier tenantIdentifier, Storage storage, LoginMethod lm)
+            throws StorageTransactionLogicException {
+        try {
+            SignInUpResponse userInfo = ThirdParty.createThirdPartyUser(
+                    tenantIdentifier, storage, lm.thirdPartyId, lm.thirdPartyUserId, lm.email,
+                    lm.timeJoinedInMSSinceEpoch);
+
+            lm.superTokensUserId = userInfo.user.getSupertokensUserId();
+        } catch (StorageQueryException | TenantOrAppNotFoundException e) {
+            throw new StorageTransactionLogicException(e);
+        } catch (DuplicateThirdPartyUserException e) {
+            throw new StorageTransactionLogicException(new Exception("A user with thirdPartyId " + lm.thirdPartyId
+                    + " and thirdPartyUserId " + lm.thirdPartyUserId + " already exists"));
+        }
+    }
+
+    private void processPasswordlessLoginMethod(TenantIdentifier tenantIdentifier, Storage storage, LoginMethod lm)
+            throws StorageTransactionLogicException {
+        try {
+            AuthRecipeUserInfo userInfo = Passwordless.createPasswordlessUser(tenantIdentifier, storage, lm.email,
+                    lm.phoneNumber, lm.timeJoinedInMSSinceEpoch);
+
+            lm.superTokensUserId = userInfo.getSupertokensUserId();
+        } catch (StorageQueryException | TenantOrAppNotFoundException | RestartFlowException e) {
```

Comment:
RestartFlowException means that this is a duplicate email and / or phone passed in. So the error needs to indicate that

Reply:
Added a custom error message for this.